File: packages/sdk/src/core/managers/session-manager.ts
import type { HttpClient } from '../client/http-client';
import type { OfflineQueue } from '../queue/offline-queue';
import type { CreateSessionRequest, PingSessionRequest } from '../types';
import { generateSessionId } from '../utils/id-generator';
import { logger } from '../utils/logger';
import { validateSessionId } from '../utils/validator';

const PING_INTERVAL_MS = 5000;
const INACTIVITY_TIMEOUT_MS = 5 * 60 * 1000;

export class SessionManager {
  private sessionId: string | null = null;
  private startPromise: Promise<string> | null = null;
  private pingInterval: ReturnType<typeof setInterval> | null = null;
  private isOnline = true;
  private pausedAt: number | null = null;
  private readonly httpClient: HttpClient;
  private readonly offlineQueue: OfflineQueue;
  private readonly deviceId: string;

  constructor(
    httpClient: HttpClient,
    offlineQueue: OfflineQueue,
    deviceId: string
  ) {
    this.httpClient = httpClient;
    this.offlineQueue = offlineQueue;
    this.deviceId = deviceId;
  }

  async start(isOnline: boolean): Promise<string> {
    if (this.startPromise) {
      return this.startPromise;
    }
    if (this.sessionId) {
      return this.sessionId;
    }

    this.startPromise = this.doStart(isOnline);

    try {
      return await this.startPromise;
    } finally {
      this.startPromise = null;
    }
  }

  private async doStart(isOnline: boolean): Promise<string> {
    this.stopPingInterval();

    this.isOnline = isOnline;
    this.sessionId = generateSessionId();

    const validation = validateSessionId(this.sessionId);
    if (!validation.success) {
      logger.error('Generated session ID invalid. Retrying.');
      this.sessionId = generateSessionId();
      const retryValidation = validateSessionId(this.sessionId);
      if (!retryValidation.success) {
        logger.error('Failed to generate valid session ID. Using fallback.');
      }
    }

    const payload: CreateSessionRequest = {
      sessionId: this.sessionId,
      deviceId: this.deviceId,
      startedAt: new Date().toISOString(),
    };

    if (isOnline) {
      const result = await this.httpClient.createSession(payload);
      if (!result.success) {
        logger.error(
          'Session creation failed. Queuing for retry.',
          result.error
        );
        await this.offlineQueue.enqueue({ type: 'session', payload });
      }
    } else {
      await this.offlineQueue.enqueue({ type: 'session', payload });
    }

    this.startPingInterval();

    return this.sessionId;
  }

  pause(): void {
    this.pausedAt = Date.now();
    this.stopPingInterval();
  }

  async resume(): Promise<void> {
    if (!this.sessionId) {
      return;
    }

    if (this.pausedAt) {
      const inactiveDuration = Date.now() - this.pausedAt;

      if (inactiveDuration > INACTIVITY_TIMEOUT_MS) {
        logger.info('Session inactive for too long, starting new session', {
          inactiveDuration: Math.round(inactiveDuration / 1000),
          threshold: Math.round(INACTIVITY_TIMEOUT_MS / 1000),
        });

        this.sessionId = null;
        this.pausedAt = null;

        await this.start(this.isOnline);
        return;
      }

      this.pausedAt = null;
    }

    this.startPingInterval();
  }

  getSessionId(): string | null {
    return this.sessionId;
  }

  updateNetworkState(isOnline: boolean): void {
    this.isOnline = isOnline;
  }

  private startPingInterval(): void {
    this.stopPingInterval();

    this.pingInterval = setInterval(() => {
      this.sendPing().catch(() => {
        logger.error('Unhandled error in sendPing. Session may timeout.');
      });
    }, PING_INTERVAL_MS);
  }

  private stopPingInterval(): void {
    if (this.pingInterval) {
      clearInterval(this.pingInterval);
      this.pingInterval = null;
    }
  }

  private async sendPing(): Promise<void> {
    if (!this.sessionId) {
      return;
    }

    const payload: PingSessionRequest = {
      sessionId: this.sessionId,
      timestamp: new Date().toISOString(),
    };

    if (this.isOnline) {
      const result = await this.httpClient.pingSession(payload);
      if (!result.success) {
        logger.error('Session ping failed. Queuing for retry.', result.error);
        await this.offlineQueue.enqueue({ type: 'ping', payload });
      }
    } else {
      await this.offlineQueue.enqueue({ type: 'ping', payload });
    }
  }
}

File: packages/swift/Sources/PhaseAnalytics/Core/Managers/SessionManager.swift
import Foundation

internal actor SessionManager {
    private var sessionID: String?
    private var pingTimer: Task<Void, Never>?
    private var isOnline = true
    private var pausedAt: Date?

    private let httpClient: HTTPClient
    private let offlineQueue: OfflineQueue
    private let deviceID: String

    private var startPromise: Task<String, Never>?

    private static let pingInterval: TimeInterval = 5.0
    private static let inactivityTimeout: TimeInterval = 5 * 60

    init(httpClient: HTTPClient, offlineQueue: OfflineQueue, deviceID: String) {
        self.httpClient = httpClient
        self.offlineQueue = offlineQueue
        self.deviceID = deviceID
    }

    func start(isOnline: Bool) async -> String {
        if let promise = startPromise {
            return await promise.value
        }

        if let existingID = sessionID {
            return existingID
        }

        let promise = Task<String, Never> {
            await doStart(isOnline: isOnline)
        }
        startPromise = promise

        let result = await promise.value
        startPromise = nil
        return result
    }

    private func doStart(isOnline: Bool) async -> String {
        stopPingInterval()

        self.isOnline = isOnline
        sessionID = IDGenerator.generateSessionID()

        if case .failure = Validator.validateSessionID(sessionID!) {
            logger.error("Generated session ID invalid, retrying")
            sessionID = IDGenerator.generateSessionID()

            if case .failure = Validator.validateSessionID(sessionID!) {
                logger.error("Failed to generate valid session ID, using UUID fallback")
                sessionID = UUID().uuidString
            }
        }

        let payload = CreateSessionRequest(
            sessionId: sessionID!,
            deviceId: deviceID,
            startedAt: currentISO8601Timestamp()
        )

        if isOnline {
            let result = await httpClient.createSession(payload)
            if case .failure = result {
                logger.error("Failed to create session, queueing")
                await offlineQueue.enqueue(.session(payload: payload, clientOrder: 0, retryCount: nil))
            }
        } else {
            await offlineQueue.enqueue(.session(payload: payload, clientOrder: 0, retryCount: nil))
        }

        startPingInterval()

        return sessionID!
    }

    func pause() {
        pausedAt = Date()
        stopPingInterval()
        logger.debug("Session paused")
    }

    func resume() async {
        guard let pausedAt = pausedAt else {
            startPingInterval()
            return
        }

        let inactiveDuration = Date().timeIntervalSince(pausedAt)

        if inactiveDuration > Self.inactivityTimeout {
            logger.info(
                "Session inactive for \(Int(inactiveDuration)) seconds, starting new session (threshold: \(Int(Self.inactivityTimeout)))"
            )

            sessionID = nil
            self.pausedAt = nil

            _ = await start(isOnline: isOnline)
        } else {
            self.pausedAt = nil
            startPingInterval()
            logger.debug("Session resumed after \(Int(inactiveDuration)) seconds")
        }
    }

    func getSessionID() -> String? {
        return sessionID
    }

    func updateNetworkState(isOnline: Bool) {
        self.isOnline = isOnline
    }

    private func startPingInterval() {
        stopPingInterval()

        pingTimer = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: UInt64(Self.pingInterval * 1_000_000_000))

                if !Task.isCancelled {
                    await sendPing()
                }
            }
        }
    }

    private func stopPingInterval() {
        pingTimer?.cancel()
        pingTimer = nil
    }

    private func sendPing() async {
        guard let sessionID = sessionID else {
            return
        }

        let payload = PingSessionRequest(
            sessionId: sessionID,
            timestamp: currentISO8601Timestamp()
        )

        if isOnline {
            let result = await httpClient.pingSession(payload)
            if case .failure = result {
                logger.error("Failed to ping session, queueing")
                await offlineQueue.enqueue(.ping(payload: payload, clientOrder: 0, retryCount: nil))
            }
        } else {
            await offlineQueue.enqueue(.ping(payload: payload, clientOrder: 0, retryCount: nil))
        }
    }
}
